### PR TITLE
feature/#764 added submission entity, manager and repo

### PR DESCRIPTION
**Pull request by toan-quach**

### CODE REVIEW

File: src/taipy/core/_orchestrator/_orchestrator.py
Comment (jrobinAV) [inline src/taipy/core/_orchestrator/_orchestrator.py]: I believe we can discuss when the submission status is updated. 
I was initially thinking of updating the status when each job status is updated.

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission.py]: Busted!

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission.py]: We need to add an alias for the submission id, like for other entities.
```suggestion
        submission_id: Optional[SubmissionId] = None,
```

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission.py:51]: We are missing the submission_status and the creation date in the signature.

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission.py]: If possible, I would call this method from the JobDispatcher in method _update_job_status.
If so, I believe all other usages are unnecessary anymore.

File: src/taipy/core/submission/_submission_converter.py
Comment (jrobinAV) [inline src/taipy/core/submission/_submission_converter.py]: Why don't you pass these in the constructor?

File: src/taipy/core/submission/_submission_manager.py
Comment (jrobinAV) [inline src/taipy/core/submission/_submission_manager.py:1]: To be completed with missing methods: etters, deleters, filters, etc...

File: src/taipy/core/submission/submission_status.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission_status.py]: todo: review

**Review by jrobinAV — COMMENTED** [review]
It looks good to me. 
It needs to be completed, in particular the unit tests.

File: src/taipy/core/_orchestrator/_orchestrator.py
Comment (toan-quach) [reply src/taipy/core/_orchestrator/_orchestrator.py]: I'm thinking we can move it to the `def status` setter in Job (which is what you proposed I believe), what do you think?

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py:51]: My thinking for it is that, we will create a Submission object before creating a job, then we will create jobs and propagate jobs to the submission object after that. So based on that flow, I think not having them in the Submission signature makes sense. But it can also be easier in different cases to have them in the signature as well, though don't think it's accurate for the use case

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py]: I think we can put it in the status setter function of Job entity

File: src/taipy/core/submission/_submission_converter.py
Comment (toan-quach) [reply src/taipy/core/submission/_submission_converter.py]: as mentioned before, I think the flow should be create a Submission entity to have a submission id before creating jobs, after that, we will create jobs with the new submission id, initially I want to make jobs to be a readable value only as well and not changeable, maybe we can discuss this more to align our thoughts :D

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [reply src/taipy/core/submission/submission.py:51]: Without the creation_date and the status in the signature, when reading a submission from our repo, we will first instantiate a submission with `status` equals to `UNDEFINED` and `creation_date` equals to `datetime.now()`. Immediately after, we will set the proper values. This is strange to me.
Plus it is not consistent with other entities.

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [reply src/taipy/core/submission/submission.py]: you are right! in status setter it is better.

File: tests/core/submission/test_submission.py
Comment (jrobinAV) [inline tests/core/submission/test_submission.py]: todo

File: src/taipy/core/_orchestrator/_orchestrator.py
Comment (jrobinAV) [inline src/taipy/core/_orchestrator/_orchestrator.py]: change name to `lock_output_and_create_job`

File: tests/core/notification/test_events_published.py
Comment (jrobinAV) [inline tests/core/notification/test_events_published.py:1]: This test is not maintainable enough. We need to split it. Let me push a proposal directly on the branch if that s fine for you.

File: tests/core/notification/test_notifier.py
Comment (jrobinAV) [inline tests/core/notification/test_notifier.py:656]: This test is not maintainable anymore. Can you split it?

File: tests/core/notification/test_events_published.py
Comment (jrobinAV) [inline tests/core/notification/test_events_published.py]: There are too many events. We publish 6 (`SUBMISSION`, `UPDATE`, `submission_status`) events while only 3 should be published.
I don't know why.

File: src/taipy/core/job/job.py
Comment (jrobinAV) [inline src/taipy/core/job/job.py]: BUG!
This does not work.
The self_setter decorator works in 3 steps: 
1. It executes the `job.status(val)` property code
2. It sets the job into the repo
3. It publishes the update event

That means in our case: 
1. So first we set the `job._status` in memory only.
    Then we call `update_submission_status` which gets the job from the repo (with the 
    previous job status value) to compute the new `submission_status` which is incorrect.
    Then we set the  INCORRECT `submission.submission_status` into the repository.
2. We set the job writing the new job status into the repository
3. We publish an update event

File: tests/core/notification/test_events_published.py
Comment (jrobinAV) [reply tests/core/notification/test_events_published.py]: I believe it comes from the auto setter decorator that execute the function twice.

File: tests/core/notification/test_events_published.py
Comment (jrobinAV) [inline tests/core/notification/test_events_published.py]: The scenario deletion should also trigger the submission deletion. We forgot to implement it.

File: tests/core/notification/test_events_published.py
Comment (jrobinAV) [inline tests/core/notification/test_events_published.py:1]: This test is False. It is hiding some bugs.

**Review by jrobinAV — CHANGES_REQUESTED** [review]

File: tests/core/notification/test_events_published.py
Comment (toan-quach) [reply tests/core/notification/test_events_published.py]: yep you're right. I've been trying to see if we can fix it but no luck so far :/

File: tests/core/notification/test_events_published.py
Comment (toan-quach) [reply tests/core/notification/test_events_published.py]: done!

File: src/taipy/core/submission/submission_status.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission_status.py:41]: Add SUBMITTED Status

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [inline src/taipy/core/submission/submission.py:144]: We should type the argument in the signature of the submission_status property setter. So we don't need the type ignore.

Comment (toan-quach) [reply src/taipy/core/submission/submission.py:144]: I believe I did but it still raises the attribute is read-only :/

File: src/taipy/core/submission/submission.py
Comment (gmarabout) [inline src/taipy/core/submission/submission.py]: 🥜 Why not simply `return self.creation_date.timestamp() <= other.creation_date.timestamp()`?

File: src/taipy/core/submission/submission.py
Comment (gmarabout) [inline src/taipy/core/submission/submission.py]: 🥜 Why not simply return self.creation_date.timestamp() >= other.creation_date.timestamp()?

File: src/taipy/core/submission/submission.py
Comment (gmarabout) [inline src/taipy/core/submission/submission.py:47]: ❓Just by curiosity, do we really need to have `Union[List[Job], List[JobId]]`? Why not a simpler `List[JobId]`? 
Having a more precise parameter type lower the risk of bugs in the long term.
M2c

File: src/taipy/core/submission/submission.py
Comment (gmarabout) [inline src/taipy/core/submission/submission.py:132]: ⚠️ This method is really hard to read and hard to understand. 
Why not something like this?
```
  if job.is_canceled():
      self.submission_status = SubmissionStatus.CANCELED
  if job.is_blocked():
      self.submission_status = SubmissionStatus.BLOCKED
  if job.is_pending() or job.is_submitted():
      self.submission_status = SubmissionStatus.PENDING
  if job.is_running():
      self.submission_status = SubmissionStatus.RUNNING
  if job.is_completed() or job.is_skipped():
      self.submission_status = SubmissionStatus.COMPLETED
  if job.is_abandoned():
      self.submission_status = SubmissionStatus.UNDEFINED
```

File: src/taipy/core/submission/submission.py
Comment (gmarabout) [inline src/taipy/core/submission/submission.py:147]: ⚠️ All the other cases have a `continue`. This one hasn't. Is it intended?

**Review by gmarabout — COMMENTED** [review]
Hey 👋🏻 
I just added some small comments on the code. 
Maybe a quick description of the PR would have helped me provide you with a better review :-)

**Review by jrobinAV — DISMISSED** [review]

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py:47]: I want to offer both possibility to pass in the Job entity or just the JobId instead of asking the user for 1 of the 2 if that makes sense to you :D since we use entity and id interchangeably, enforcing just one might be going away from the overall behavior. Though it probably still makes sense in this context as Submission is like for internal use only (at least for now)

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py:147]: haha thanks for pointing it out, I think we don't need any of the "continue"s after all!

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py:132]: I believe that's so we can loop through all of the jobs and prioritize some of the statuses. Like for example, if we have a list of canceled jobs, but 1 of them is a failed job, we would want to prioritize the failing status instead of the cancel one. Hope it make sense! 

@jrobinAV can you help fill in if I missed any detail?

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py]: yeah... my bad that was weird...

File: src/taipy/core/submission/submission.py
Comment (toan-quach) [reply src/taipy/core/submission/submission.py]: thanks for pointing this out! I noticed the same issue on Job, will be fixing that as well!

File: src/taipy/core/submission/submission.py
Comment (gmarabout) [reply src/taipy/core/submission/submission.py:47]: I leave you the final decision of course, and it's not a big deal anyway :-)

But as a user, I think I would prefer clear and explicit API. 

With the one you propose, the caller may wonder if passing IDs or Jobs would make the function work differently. 
For example, would the function avoid any additional data loading if we pass the job?
Again, not really a pb, and we can deal with this API :-)

Comment (jrobinAV) [reply src/taipy/core/submission/submission.py:47]: If we want to change this convention, or at least discuss it, it should be a global decision. It is out of this PR scope.
My2cents

Comment (toan-quach) [reply src/taipy/core/submission/submission.py:47]: I agree :D as the behavior is quite global, it's not the best place to discuss it!

File: src/taipy/core/submission/submission.py
Comment (jrobinAV) [reply src/taipy/core/submission/submission.py:132]: I don't know what to say. 😄 
The behavior of your code proposal @gmarabout, is not the same as what we implemented. 
It does not respect what we want to implement.
We have to parse all the jobs before computing the submission_status.

Comment (gmarabout) [reply src/taipy/core/submission/submission.py:132]: > It does not respect what we want to implement.

That's is my point: the intent of the method is not obvious when you read it.

Comment (toan-quach) [reply src/taipy/core/submission/submission.py:132]: @gmarabout 
 would renaming it to `_update_submission_status_based_on_jobs` help?

Comment (gmarabout) [reply src/taipy/core/submission/submission.py:132]: Definitely not :-) 
The name should reflect the intent of the method, what it is supposed to do.
Not *how* it does it. 
What if we decide to change the logic in the future?
The code should definitely speak by itself about the *how*...
If you think the code is clear enough, then let's go. I don't want to block this PR.

Comment (toan-quach) [reply src/taipy/core/submission/submission.py:132]: alright thanks Gregoire! I'll merge it for now and think how i can improve it later

**Review by jrobinAV — APPROVED** [review]

**Pull request merged**
